presentation/backend/src/presentation-backend/ElementPropertiesHelper.ts

```diff
@@ -37,70 +34,285 @@ function getECSqlName(fullClassName: string) {
+      mergeMap(
+        ({ classFullName, ids }) =>
+          getBatchedClassContentItems(
+            classFullName,
+            contentDescriptorGetter,
+            contentSetGetter,
+            () => createIdBatches(OrderedId64Iterable.sortArray(ids), batchSize),
+            batchesParallelism,
+          ),
+        classParallelism,
+      ),
+    ),
+    count: of(elementIds.length),
+  };
 }
 
 /** @internal */
-export async function getBatchedClassElementIds(
+export function getContentItemsObservableFromClassNames(
   imodel: IModelDb,
-  fullClassName: string,
+  contentDescriptorGetter: (
+    partialProps: Pick<ContentDescriptorRequestOptions<IModelDb, KeySet, RulesetVariable>, "rulesetOrId" | "keys">,
+  ) => Promise<Descriptor | undefined>,
+  contentSetGetter: (
+    partialProps: Pick<ContentRequestOptions<IModelDb, Descriptor, KeySet, RulesetVariable>, "rulesetOrId" | "keys" | "descriptor">,
+  ) => Promise<Item[]>,
+  elementClasses: string[],
+  classParallelism: number,
+  batchesParallelism: number,
   batchSize: number,
-): Promise<Array<{ from: Id64String; to: Id64String }>> {
-  const batches = [];
-  const reader = imodel.createQueryReader(`SELECT ECInstanceId id FROM ${getECSqlName(fullClassName)} ORDER BY ECInstanceId`);
-  let currId: Id64String | undefined;
-  let fromId: Id64String | undefined;
-  let count = 0;
-  while (await reader.step()) {
-    currId = reader.current.toRow().id as Id64String;
-    if (!fromId) {
-      fromId = currId;
-    }
-    if (++count >= batchSize) {
-      batches.push({ from: fromId, to: currId });
-      fromId = undefined;
-      count = 0;
+): { itemBatches: Observable<{ descriptor: Descriptor; items: Item[] }>; count: Observable<number> } {
+  return {
+    itemBatches: getClassesWithInstances(imodel, elementClasses).pipe(
+      mergeMap(
+        (classFullName) =>
+          getBatchedClassContentItems(
+            classFullName,
+            contentDescriptorGetter,
+            contentSetGetter,
+            () => getBatchedClassElementIds(imodel, classFullName, batchSize),
+            batchesParallelism,
+          ),
+        classParallelism,
+      ),
+    ),
+    count: of(getElementsCount(imodel, elementClasses)),
+  };
+}
+
+function getBatchedClassContentItems(
+  classFullName: string,
+  contentDescriptorGetter: (
+    partialProps: Pick<ContentDescriptorRequestOptions<IModelDb, KeySet, RulesetVariable>, "rulesetOrId" | "keys">,
+  ) => Promise<Descriptor | undefined>,
+  contentSetGetter: (
+    partialProps: Pick<ContentRequestOptions<IModelDb, Descriptor, KeySet, RulesetVariable>, "rulesetOrId" | "keys" | "descriptor">,
+  ) => Promise<Item[]>,
+  batcher: () => Observable<Array<{ from: Id64String; to: Id64String }>>,
+  batchesParallelism: number,
+): Observable<{ descriptor: Descriptor; items: Item[] }> {
+  return of({
+    ruleset: createClassContentRuleset(classFullName),
+    keys: new KeySet(),
+  }).pipe(
+    mergeMap(({ ruleset, keys }) =>
```

Comment:
Looks like this can be simplified by removing `of` and first level of `mergeMap` and instead returning `defer` that created ruleset and keys inside?

Reply:
Changed.

Reply:
My idea was to have only single `defer` as I am not sure why multiple are needed:

```ts
return defer(async () => {
  const ruleset = createClassContentRuleset(classFullName);
  const keys = new KeySet();
  const descriptor = await contentDescriptorGetter({ rulesetOrId: ruleset, keys });
  if (!descriptor) {
    throw new PresentationError(PresentationStatus.Error, `Failed to get descriptor for class ${classFullName}`);
  }
  return descriptor;
}).pipe(
```

Reply:
Changed.